Enforce context deadline when fetching apps or sessions

adshmh opened the pull request:
## Summary

When fetching a single app, all apps, or a session, the supplied context's deadline (if any) is enforced:
`GetApplication`, `GetAllApplications`, and `GetSession` will return an error as soon as the supplied context's deadline is exceeded, without waiting for the full node's response.

## Issue

Without this fix, users of the SDK need to write custom code for forcing the desired deadline, even though it is passed to the SDK in a `context.Context`.

- #{ISSUE_NUMBER}

## Type of change

Select one or more:

- [x] New feature, functionality or library
- [ ] Bug fix
- [ ] Code health or cleanup
- [ ] Documentation
- [ ] Other (specify)

## Sanity Checklist

- [x] I have tested my changes using the available tooling
- [x] I have commented my code
- [x] I have performed a self-review of my own code; both comments & source code
- [ ] I create and reference any new tickets, if applicable
- [ ] I have left TODOs throughout the codebase, if applicable

Olshansk commented on session.go:
```diff
@@ -30,27 +30,49 @@ func (s *SessionClient) GetSession(
+		// TODO_TECHDEBT(@adshmh): consider increasing the default response size:
+		// e.g. using google.golang.org/grpc's MaxCallRecvMsgSize CallOption.
+		//
+		// TODO_IMPROVE: Would it be feasible to add a GetCurrentSession, supported by the underlying protocol?
```
https://github.com/pokt-network/poktroll/blob/6afdff0b0d41c49677e62edad2af3de5c2506cc2/x/session/keeper/query_get_session.go#L16

<img width="883" alt="Screenshot 2025-04-16 at 1 59 43 PM" src="https://github.com/user-attachments/assets/91f13f85-8da9-4379-a17b-8eee12d33ddc" />

1. Can you try this?
2. If it works -> remove the TODO
3. If it doesn't work -> Please open an issue in ght poktroll repo because it's a bug

adshmh replied:
Thank you for the review. This TODO was just moved around, not added.
It is now removed. This is already how PATH uses the SDK:

https://github.com/buildwithgrove/path/blob/4d94e2e6bfea5a4774c3a7263c0cce69338d95b4/protocol/shannon/fullnode_lazy.go#L94-L100

![image](https://github.com/user-attachments/assets/3ecf17e5-67ba-4c1b-8b43-3972aa89f4e6)

Olshansk reviewed (changes requested):
@adshmh Requesting changes becase the TODO is arguably incorrect depending on the results of your future test/verification.

adshmh commented on application.go:
```diff
@@ -39,32 +39,74 @@ type ApplicationClient struct {
-	if err != nil {
-		return []types.Application{}, err
-	}
+	resultCh := make(chan result, 1)
```
This has been addressed through a simplification refactor.

adshmh commented on application.go:
```diff
@@ -39,32 +39,74 @@ type ApplicationClient struct {
-		return types.Application{}, err
+	// Enforce any deadlines specified on the supplied context.
+	// Create channel for result
+	type result struct {
```
This has been addressed through a simplification refactor.

red-0ne reviewed (approved):
Much simpler! :+1:

Olshansk commented:
@adshmh Please make sure to update the shannon-sdk sha in the `path` and `poktroll` repos.

No need for a review / approval. Just self-approve your own PR.

adshmh commented:
> @adshmh Please make sure to update the shannon-sdk sha in the `path` and `poktroll` repos.
> 

- Updated on `path`:

https://github.com/buildwithgrove/path/blob/e116af8d3016ef65b991d99ed06b92b4c75094ce/go.mod#L18

Was done through PR 213:
https://github.com/buildwithgrove/path/pull/213/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6

- Updated on `poktroll`:
https://github.com/pokt-network/poktroll/blob/b92aa0c6e01b300825a25e4121bd8279b6cdcf53/go.mod#L60

through PR 1241:
https://github.com/pokt-network/poktroll/pull/1241
